components/feature/tabs/src/main/java/mozilla/components/feature/tabs/WindowFeature.kt

```diff
@@ -0,0 +1,82 @@
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.mapNotNull
+import mozilla.components.browser.state.action.ContentAction
+import mozilla.components.browser.state.selector.selectedTab
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.concept.engine.window.WindowRequest
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+
+/**
+ * Feature implementation for handling window requests by opening and closing tabs.
+ */
+class WindowFeature(
+    private val store: BrowserStore,
+    private val tabsUseCases: TabsUseCases
+) : LifecycleAwareFeature {
+
+    private var scope: CoroutineScope? = null
+
+    /**
+     * Starts observing the selected session to listen for window requests
+     * and opens / closes tabs as needed.
+     */
+    override fun start() {
+        scope = store.flowScoped { flow ->
+            flow.mapNotNull { state -> state.selectedTab }
+                .ifChanged {
+                    it.content.windowRequest
+                }
+                .collect { state ->
+                    val windowRequest = state.content.windowRequest
+                    when (windowRequest?.type) {
+                        WindowRequest.Type.CLOSE -> consumeWindowRequest(state.id) {
+                            tabsUseCases.removeTab(state.id)
+                        }
+                        WindowRequest.Type.OPEN -> consumeWindowRequest(state.id) {
+                            if (state.content.private) {
+                                tabsUseCases.addPrivateTab(
+                                    url = windowRequest.url,
+                                    selectTab = true,
+                                    parentId = state.id,
+                                    engineSession = windowRequest.prepare()
+                                )
+                            } else {
+                                tabsUseCases.addTab(
+                                    url = windowRequest.url,
+                                    selectTab = true,
+                                    parentId = state.id,
+                                    engineSession = windowRequest.prepare()
+                                )
+                            }
+                            windowRequest.start()
+                        }
+                    }
+                }
+        }
+    }
+
+    private fun consumeWindowRequest(
+        tabId: String,
+        consume: () -> Unit
+    ) {
+        consume()
+        store.dispatch(ContentAction.ConsumeWindowRequestAction(tabId))
+    }
+
+    /**
+     * Stops observing the selected session for incoming window requests.
+     */
+    override fun stop() {
+        scope?.cancel()
```

Comment:
`flowScoped` returns `MainScope`, which `has [SupervisorJob] and [Dispatchers.Main] context elements.`.

What effect will cancellation have?

Reply:
Ah yes, this took me a while to get used to as well! I also assumed that `MainScope().cancel` would be a problem, because I misinterpreted what it was doing. This is really just cancelling the scope we create here: https://github.com/mozilla-mobile/android-components/blob/master/components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/PromptRequest.kt

So, we really just want to stop observing / consuming from the flow, until the feature is started again. It doesn't stop / cancel anything else.

Reply:
You linked a wrong url, right? Or i'm just not seeing it.

But OK, I understand it now. This is just cancelling a just-created `SupervisorJob`. Well, this takes some mental gymnastics to read... Maybe `MainScope` could be named differently, like... `JobOnMainDispatcher`? :)

Reply:
Ah yeah wrong URL, meant to link to https://github.com/mozilla-mobile/android-components/blob/master/components/lib/state/src/main/java/mozilla/components/lib/state/ext/StoreExtensions.kt#L182

Yes, the naming isn't ideal, maybe we should create an alias and not use the lib function.